### PR TITLE
[Fix #2408] Alternative setTimeout for long timers

### LINKED ISSUE #2408: Find an alternative to setTimeout for long timers
#In the latest version of react-native there is a warning in Status regarding setTimeout:

> Setting a timer for a long period of time, i.e. multiple minutes, is a performance and correctness issue on Android as it keeps the timer module awake, amd timers can only be called when the app is in the foreground. See https://github.com/facebook/react-native/issues/12981 for more info.
> (Saw setTimeout with duration 180000ms)

As far as I know this only concerns this line of code:
https://github.com/status-im/status-react/blob/edd4df85f95b42a459b9b189918c3b3f76102e3f/src/status_im/ui/screens/discover/events.cljs#L80-L82

`dispatch-later` is not a viable alternative because it uses setTimeout internally as well.

### PR DESCRIPTION
**Pull request by foopang**
fixes #2408

First I tried replacing `setTimeout` `setInterval` globally as suggested in https://github.com/facebook/react-native/issues/12981#issuecomment-332697960, but it didn't work unfortunately. So what I did in this PR is find all `setTimeout` `setInterval` and replace them with the background timer version of `setTimeout` and `setInterval`. 

Errors show below when set globally to replace `setTimeout` and `setInterval`.

```
TypeError: expected dynamic type `double', but had type `null' (constructing arguments for RNBackgroundTimer.setTimeout at argument index 1)

TypeError: expected dynamic type `double', but had type `null'
invoke
    JavaMethodWrapper.java:368
invoke
    JavaModuleWrapper.java:162
run
    NativeRunnable.java
handleCallback
    Handler.java:751
dispatchMessage
    Handler.java:95
dispatchMessage
    MessageQueueThreadHandler.java:31
loop
    Looper.java:154
run
    MessageQueueThreadImpl.java:194
run
    Thread.java:761
```

status: ready

### CODE REVIEW

**Comment by oskarth** [comment]
Requesting 3 reviewers based on recently edited files.

File: ios/Podfile [hunk 1]
```diff
@@ -18,7 +18,11 @@ target 'StatusIm' do
   # Pods for StatusIm
   pod 'Instabug', '~> 7.0'
   pod 'FirebaseMessaging'
-  
+
+  pod 'React', :path => '../node_modules/react-native'
+  pod 'yoga', :path => '../node_modules/react-native/ReactCommon/yoga'
```
Comment (jeluard) [inline ios/Podfile:23]: Can you detail why those changes?

File: ios/Podfile.lock [hunk 2]
```diff
@@ -20,10 +20,28 @@ PODS:
     - GoogleToolboxForMac/Defines (= 2.1.1)
   - Instabug (7.2.6)
   - Protobuf (3.3.0)
+  - React (0.50.4):
+    - React/Core (= 0.50.4)
+  - react-native-background-timer (2.0.0):
```
Comment (jeluard) [inline ios/Podfile.lock]: Please comment here too.

File: src/status_im/utils/utils.cljs [hunk 3]
```diff
@@ -140,3 +140,24 @@
...
+(defn clear-timeout [id]
+  (.clearTimeout rn-dependencies/background-timer id))
+
+(defn set-interval [cb ms]
+  (.setInterval rn-dependencies/background-timer cb ms))
+
+(defn clear-interval [id]
+  (.clearInterval rn-dependencies/background-timer id))
+
+;; replace globally
```
Comment (jeluard) [inline src/status_im/utils/utils.cljs]: Remove as it is unused

**Review by jeluard — COMMENTED** [review]
Do we want to replace all timeouts or only the long ones?

File: ios/Podfile (hunk 1 above)
Comment (foopang) [reply ios/Podfile:23]: It was added automatically when I ran `react-native link` after adding `react-native-background-timer`. I believe it's required by `React/Core` as a dependency. And `react-native-background-timer` needs `React` as a dependency, so it ended up having to add `React` > `React/Core` > `yoga`.

File: ios/Podfile.lock (hunk 2 above)
Comment (foopang) [reply ios/Podfile.lock]: Same as above.

**Comment by foopang** [comment]
@jeluard quoted from https://github.com/ocetnik/react-native-background-timer#usage-android

> You can use the setInterval and setTimeout functions. This API is identical to that of react-native and can be used to quickly replace existing timers with background timers.

So I think it's safe to replace all timeouts. Is there any concern you have?

**Review by jeluard — APPROVED** [review]
Looks good to me!

**Review by oskarth — APPROVED** [review]

**Comment by oskarth** [comment]
Moving to TO TEST to rule out https://github.com/status-im/status-react/issues/2990 being related to this issue

**Comment by oskarth** [comment]
@foopang can you please resolve conflicts btw?

**Comment by foopang** [comment]
@oskarth conflicts resolved.

**Comment by asemiankevich** [comment]
@foopang @oskarth I can't build it , maybe rebase or something?

**Comment by foopang** [comment]
@asemiankevich Rebased. Let me know if you can build it now.

**Comment by asemiankevich** [comment]
@foopang build failed again
`Testing status-im.test.utils.async
undefined
/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/utils/utils.cljs:147
  (.setTimeout rn-dependencies/background-timer cb ms))
               ^
TypeError: Cannot read property 'setTimeout' of undefined
    at status_im$utils$utils$set_timeout (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/utils/utils.cljs:147:16)
    at status_im$utils$async$timeout (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/utils/async.cljs:8:3)
    at /Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/native_module/impl/module.cljs:42:3
    at /Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/native_module/impl/module.cljs:42:3
    at status_im$native_module$impl$module$state_machine__24116__auto____1 (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/native_module/impl/module.js:297:4)
    at status_im$native_module$impl$module$state_machine__24116__auto__ (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/native_module/impl/module.js:313:76)
    at cljs$core$async$impl$ioc_helpers$run_state_machine (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/cljs/core/async/impl/ioc_helpers.cljs:35:23)
    at cljs$core$async$impl$ioc_helpers$run_state_machine_wrapped (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/cljs/core/async/impl/ioc_helpers.cljs:39:6)
    at /Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/status_im/native_module/impl/module.cljs:42:3
    at Immediate.cljs$core$async$impl$dispatch$process_messages [as _onImmediate] (/Users/jenkins/workspace/react_pull_requests_PR-2853-RGLF332JGKAMJT2J4RZCIXHCOWDWWPYLLWRTCNBVSBPTYETWKF7A/target/test/cljs/core/async/impl/dispatch.cljs:18:7)
Error encountered performing task 'doo' with profile(s): 'test'
Subprocess failed`

**Comment by asemiankevich** [comment]
----------------
Branch: PR-2853
Android: https://i.diawi.com/UCgkwj
iOS: https://i.diawi.com/wwGgXi

the following scenario is tested and passed.
- add a contact on device A
- turn off the app on device A
- add a status on the contact on device B
- turn off the app on device B
- login on device A
- login the contact on other device B
- status of B should appear on A after exactly 10 min

**Comment by oskarth** [comment]
> Moving to TO TEST to rule out #2990 being related to this issue

It doesn't appear to be related to #2990 and @yenda had some thoughts on short vs long timers (which I agree with), so moving it back to REVIEW to get a few more thoughts. Sorry @foopang!

**Comment by oskarth** [comment]
Please review @yenda, @rasom and @janherich (or 2/3 at least)

Also @jeluard your review isn't showing up for some reason.

**Comment by janherich** [comment]
@foopang Can you please rebase on newest develop ?

**Comment by foopang** [comment]
@janherich rebased.

@oskarth @yenda do you prefer not to use this background timer at all for ios? As those warnings only show in android. Although It is said that ios doesn't allow long background timer, I haven't yet encountered any issues with the background timer while testing with ios simulator.

**Review by janherich — APPROVED** [review]

**Comment by yenda** [comment]
@foopang I think there is no issue technically with background timers on iOS, but using them for expensive stuff gets you refused on the app store.

https://developer.apple.com/library/content/documentation/iPhone/Conceptual/iPhoneOSProgrammingGuide/BackgroundExecution/BackgroundExecution.html
```
For tasks that require more execution time to implement, you must request specific permissions to run them in the background without their being suspended. In iOS, only specific app types are allowed to run in the background:

Apps that play audible content to the user while in the background, such as a music player app
Apps that record audio content while in the background
Apps that keep users informed of their location at all times, such as a navigation app
Apps that support Voice over Internet Protocol (VoIP)
Apps that need to download and process new content regularly
Apps that receive regular updates from external accessories
```

**Comment by foopang** [comment]
@yenda Thanks for the clarification.

Actually still getting warnings if we remove "Setting a timer" from "ignoredYellowBox", I think it is because of re-frame `:dispatch-later` we are using. Should we make `:dispatch-later` our own version that could make use of the background timer internally?

**Comment by yenda** [comment]
@foopang This is probably the `dispatch-later` effect in `discover.event` ns, it is the only one that is long enough to trigger the warning afaik. You can replace it by a set-interval if you move the dispatch in the `:init-discoveries event`

**Comment by foopang** [comment]
@yenda Updated. Let me know if I have done it right https://github.com/status-im/status-react/pull/2853/commits/1c67b82b32bdfd466b71280537aa5e0403169aa4

**Review by rasom — APPROVED** [review]

**Comment by oskarth** [comment]
@foopang Please resolve conflicts.

**Comment by foopang** [comment]
@oskarth Conflicts resolved now.

**Comment by oskarth** [comment]
Thanks @foopang!

**Pull request merged**